test/typeEvaluate.test.ts

```diff
@@ -2406,6 +2430,800 @@ t.test('function: math::*', (t) => {
+              type: 'string',
+            },
+          },
+        },
+      },
+    ),
+  },
+
+  {
+    name: 'Test that order matters and attributes overrides conditionals',
+    attributes: [
+      {
+        type: 'ObjectConditionalSplat',
+        condition: {type: 'Parameter', name: 'someParam'},
+        value: nodeWithType({
+          type: 'object',
+          attributes: {
+            foo: {type: 'objectAttribute', value: {type: 'string'}},
+          },
+        }),
+      },
+      {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'string'})},
+    ],
+    expects: {
+      type: 'object',
+      attributes: {
+        foo: {
+          type: 'objectAttribute',
+          value: {
+            type: 'string',
+          },
+        },
+      },
+    },
+  },
+
+  {
+    name: 'Test with attribute and conditional splat with condition being a parameter',
+    attributes: [
+      {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'string'})},
+      {
+        type: 'ObjectConditionalSplat',
+        condition: {type: 'Parameter', name: 'someParam'},
+        value: nodeWithType({
+          type: 'object',
+          attributes: {
+            bar: {type: 'objectAttribute', value: {type: 'string'}},
+          },
+        }),
+      },
+    ],
+    expects: unionOf(
+      {
+        type: 'object',
+        attributes: {
+          foo: {
+            type: 'objectAttribute',
+            value: {
+              type: 'string',
+            },
+          },
+          bar: {
+            type: 'objectAttribute',
+            value: {
+              type: 'string',
+            },
+            optional: true,
+          },
+        },
+        rest: undefined,
+      },
+      {
+        type: 'object',
+        attributes: {
+          foo: {
+            type: 'objectAttribute',
+            value: {
+              type: 'string',
+            },
+          },
+        },
+      },
+    ),
+  },
+
+  {
+    name: "Test with a type we can't splat over(array)",
+    attributes: [
+      {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'boolean'})},
+      {
+        type: 'ObjectConditionalSplat',
+        condition: {type: 'Parameter', name: 'someParam'},
+        value: nodeWithType({
+          type: 'array',
+          of: unionOf(
+            {
+              type: 'object',
+              attributes: {
+                bar: {type: 'objectAttribute', value: {type: 'string'}},
+              },
+            },
+            {
+              type: 'object',
+              attributes: {
+                baz: {type: 'objectAttribute', value: {type: 'string'}},
+              },
+            },
+          ),
+        }),
+      },
+    ],
+    expects: unionOf(
+      {
+        type: 'object',
+        attributes: {
+          foo: {type: 'objectAttribute', value: {type: 'boolean'}},
+        },
+      },
+      {
+        type: 'unknown',
+      },
+    ),
```

Comment:
We could short circuit so that if any of the returned elements are unknown we return unknown. Guess this could happen if you have a schema where a field is either an object or and array of something, not sure that's very realistic, though it could be during a migration for example. 

If you have multiple attributes in the object, you could simplify the type check by just asserting one value and typescript would assume it's not the unknown object 🤔 

Personally no strong opinions